### DISCUSSION TITLE
Elysia 1.0 feedback

### DISCUSSION THREAD

**Discussion by SaltyAom** [elysiajs/elysia#642] [Ideas]
We would like to hear your thoughts, and feedback on your experience with Elysia 1.0

If you would like some ideas, here are some examples:
- How was your overall experience?
- What do you think is missing?
- Any pain points you are having?

Your feedback and ideas will be used to improve Elysia and the development of Elysia 1.1

Thank you

**Comment by mrctrifork** [comment]
Honestly, it's been AMAZING, and it has become my favourite and goto webserver in TS. 

* If I have to bring something up I'd mention that I am missing a proper `@elysiajs/logging` library that would log incoming requests and time spent on those automatically.

* I'd also mention that some mechanism to properly unit test websockets without ".listen()" on the tests. 

* Finally; I believe the way DI is handled can be improved as it feels like there is a lot of friction in case you have isolated modules and they are not "properly chained" or; to some comments I've read; feels very counter-intuitive.

(and some of the TypeScript errors on Eden side aren't the clearest and most of the times is just that they have mismatching versions of elysia)

Otherwise, brilliant. I love it and I look forward to keep contributing to elysia and other packages. Thank you so much for this incredible tool.


<img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZ2p1OXFjNTlxcDJ1b2Y2NDh5bm93MW1kemlnM2hpcWxpa2c3a3VwMiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xuK0uWYApXbE6mMODf/giphy.gif">

**Comment by kravetsone** [comment]
It is possible to inject types by using the macro api? it would be great...

**Comment by ajoslin** [comment]
- How was your overall experience?

Great, using Elysia+Eden+Swagger for full stack API sharing on some new services. Really love the speed, both speed of DX and actual runtime performance.

- What do you think is missing?

zod integration would be nice, not typebox only. Am using zod elsewhere in the stack.

- Any pain points you are having?

Query / body validation errors are super cryptic and don't tell you what's wrong, only that validation failed. Doesn't help you out much there.

Also, not sure how error handling should be done, no docs on way to tell if error is instanceof elysia error or not so just duck type checking it in the onError handler.

**Reply by pauek** [reply]
Zod +1

**Reply by ghost** [reply]
> Query / body validation errors are super cryptic and don't tell you what's wrong, only that validation failed. Doesn't help you out much there.

Yes. I have to do something like this:
```ts
if (code === 'VALIDATION') {
    set.status = 400
    const err = error.validator.Errors(error.value).First()
    if (err.type === 46 && err.path === '') message = 'Bad body'
    else message = err.schema.error || 'Unknown error'
}
```

**Comment by ThatOneCalculator** [comment]
> How was your overall experience?

Pretty damn good! Eden Treaty is the biggest draw for me 

> What do you think is missing?

Automatic file based routing (preferably first party) that also works properly with Eden

> Any pain points you are having?

Some of the design patterns are a bit strange, having a solid example of controller based routing would be helpful. Also GET with optimal parameters, but that's coming as of 1.1.0 so :P

**Comment by indrazm** [comment]
Hey There.
I am currently building a tech stack including Elysiajs. 
https://bexlite.dev

I have use Elysia mostly with HTMX, both in dev and production.
Everything work super nice!
Here is my feedback.

_How was your overall experience?_

Everything works seamlessly.
Fast and Easy. 5 out of 5 stars!

_What do you think is missing?_

Nothing.

_Any pain points you are having?_

I think couple of Types. For example if i need to pass JWT types into controllers,
I need to rewrite the Types. it would be great if i can just extend from.

**Thanks for building this tool!**

**Comment by alumni** [comment]
I haven't spend that much time with Elysia, so my understanding is quite limited.

For very small projects, the express-style API might be enough, but as soon as you add more than just a few routes (or the routes have complex implementations), you can easily end up creating a code mess.

So I'm quite happy to see that MVC is supported, it could make the code more SOLID. I'm still not 100% sure how to use it based on the docs, or if I can achieve the same things I can with NestJS, but I'm enthusiastic about it :) I like the class-based approach of NestJS to define controllers and services. What I like the most about NestJS though is it's dependency injection capabilities, no other JS DI library comes close to it.

What's not clear to me is how I would integrate Elysia's models with some ORM framework, e.g. TypeORM. Since I like the class-based approach for controllers, probably I'd also like models to be classes.

I think I like Elysia's concept of a context and request dependent services. I'm not really sure how it would work when you have services depending on other services. One of the pain points in NestJS is that you are allowed to inject the request into your services, which re-creates the whole DI tree at every request (unless you think about this and implement something like AsyncLocalStorage).

**Comment by WedersonCD** [comment]
I tried to convert my current project to Elysia, but the idea of putting my routes, controllers, and type validation in a single file with a unified chain of calls is really insane for me.

Maybe I was structuring my project wrong due to my lack of knowledge, but when I finished converting my user routes, I started to lose productivity trying to find where some route code starts and ends. (The same for the type -checking definitions.

My project got faster in running and responding to the requests, but the loss of readability and maintainability in the code made me roll back to Express.

**Reply by alumni** [reply]
What's the difference between Elysia and Express in the way you structure code?

I've always considered expressjs a nightmare to work with (compared to NestJs) because of the reasons you mention here :p

Express also tends to create messy code with its functional way of handling routes. So I tend to avoid this architectural style for mid-to-large sized projects.

**Reply by WedersonCD** [reply]
The main difference is in the recomendation at the image bellow.

In express I can easily navegate in my code and know what function is handle each request, because my routing is separeted from my controllers.

In Elysia I have to hit ctrl+f to find my route with a not clear view of where my controller function ends and starts the others.

I didn't know this 'NestJs' I will look for it.

<img width="384" alt="image" src="https://github.com/elysiajs/elysia/assets/28493584/34196d6c-2d2e-4817-8a85-b858760e45c4">
<img width="207" alt="image" src="https://github.com/elysiajs/elysia/assets/28493584/6306e09c-778a-4a11-82da-59f451ddec61">

**Reply by dangdennis** [reply]
Probably because Elysia wants to keep their examples simple.

But they have examples where you .use other elysia instances that form their own groups of routes. I'm doing that myself.

It is a bit annoying to have to spend brain power to decide but it favors quick starts.

**Reply by WedersonCD** [reply]
I used the .use for separated routes.

But the single '/users' routes and subrotes are big itself to create a big hard to read file.

**Comment by MAST1999** [comment]
OK, here it goes.

The experience has been pretty smooth, but I must say we are using Node with Elysia. Better integration with Node would be nice.
We tried to use Bun but hit too many blockers in our monorepo.

Other than that, it would be nice to have an api that would enable validating types without it throwing. Something like zod's `safeParse`.
I couldn't find anything in the typebox that would enable this, and didn't have time to look in to making a wrapper.

My current use case includes aggregating data from multiple servers and then serving those, so I use a lot of custom typebox validations using `Value.Decode` but it always throws. As I said above, a `safeParse` equivalent would make my job so much easier.

Also, I have a lot of string enums so including one in Elysia would be nice.

But in general it has been a breeze to work with, and I enjoy it quite a lot.

One last thing that would be nice is if it was possible to navigate from eden api to the Elysia api definition, similar to trpc's goto definition.

**Reply by crishoj** [reply]
Perhaps this pattern is what you're looking for?

```ts
const checker = TypeCompiler.Compile(schema)
checker.Check(payload)  # → boolean
checker.Errors(payload) # → ValueErrorIterator
```

**Reply by MAST1999** [reply]
Not exactly.

I'm looking for something like this:
```ts
const MyType = Type.Object({ id: Type.String() });
const validation = Value.SafeDecode(MyType, { id: 'hello' });
if (validation.error) {
	return error('Bad Request', validation.error);
}

const data = validation.data;
```

**Comment by crishoj** [comment]
I'd like some _idiomatic_ way to perform custom validation, for cases where a static schema isn't feasible.

Consider a route like this:

```ts
new Elysia()
  .post('/images/:id/crop', () => ..., {
        query: t.Object({
            height: t.Numeric({minimum: 1, maximum: ?}),
            width: t.Numeric({minimum: 1, maximum: ?})
        })
    })
```

The maximum allowed value for `height` and `width` is not known ahead of time — it depends on the particular image ID passed in as route parameter.

Obviously, custom validation can be achieved with `beforeHandle`, but care must be taken to match the regular response format for validation errors. Perhaps some kind of `validation` hook, where the developer can return structured errors?

**Comment by MAST1999** [comment]
I also just saw ArkType announce V2 which is a lot faster than zod so I'm wondering if it's fast enough to be integerated in to Elysia:
https://twitter.com/arktypeio/status/1791413688676544545

**Reply by re-thc** [reply]
Typebox that Elysia uses is faster than ArkType.

**Comment by rayriffy** [comment]
It would be nice if Elysia also expose current version that running on, it will have a good usage for plugin maintainers to control which version of Elysia is compatible with the plugin before actually running.

```ts
export const plugin = () => {
  return (app: Elysia) => {
    console.log(app.runtime) // 1.0.21 
  }
}
```

**Comment by directormac** [comment]
Would love to also get the `.subscribe()` api for server sent events also like on the websocket api related to eden treaty.
https://elysiajs.com/eden/treaty/websocket.html

**Comment by raix** [comment]
First off: thank you for open sourcing this project! ❤️ 

I'm on the looks for an Api framework with TypeScript client and Swagger support - so currently testing out Elysia _(migrating from Express+trpc)_

Feedback from an "Elysia noob" (me):

* Ergonomics of the code - perfect!
* TypeScript support - awesome!
* Eden Api client - awesome!
* Performance - solid!
* Open Api generation and UI - thank you!
* OTEL - thank you!

Not really on Elysia - something the community can help out on maybe?
* Plugins - early days - not sure if it would be possible to have native support for "Connect" type middlewares (and tap into existing OSS communities)
* Have Elysia act as an Express/Connect plugin for incremental adoption?
* Zod - TypeBox is sound decision, but would love Zod support _(already using Zod trying to minimize the mental context)_
* Docs - Missing a "Express to Elysia migration" _(took some effort to figure out how to "map" between things)_
* Bun - Still early days, would be great with an escape hatch for Node js
* XP - Production ready / large code bases?

Thanks again, take the feedback with a grain of salt - it's a five star library ⭐⭐⭐⭐

Kind regards Morten

**Reply by kravetsone** [reply]
> Plugins - early days - not sure if it would be possible to have native support for "Connect" type middlewares (and tap into existing OSS communities)

My library - [elysia-connect-middleware](https://github.com/kravetsone/elysia-connect-middleware)

**Comment by jonyw4** [comment]
Zod integration would be awesome!
